adds quantized implementation of hard sigmoid

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34607 adds quantized implementation of hard sigmoid**

Summary:

Adds quantized version of hardsigmoid activation.

Note: not implementing the _ and .out versions is
currently intended, because the implementation changes the scale and
zp and it's nice to not allow the user to specify scale
and zp.  Lmk if we should handle this differently.

Test Plan:

tests
benchmarks

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20480546](https://our.internmc.facebook.com/intern/diff/D20480546)

vkuzo committed:
adds quantized implementation of hard sigmoid

Summary:

Adds quantized version of hardsigmoid activation.

Note: not implementing the _ and .out versions is
currently intended, because the implementation changes the scale and
zp and it's nice to not allow the user to specify scale
and zp.  Lmk if we should handle this differently.

Test Plan:

tests
benchmarks

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -306,6 +306,69 @@ void qsigmoid_kernel(const Tensor& qx, Tensor& qy) {
+
+    // Naive implemenentation: uses dequantize/execute/quantize routine
+    // - Output scale is set to 1.0 / 2^(BIT_NUM)
+    // - For signed types output zero point is set to 0
```
I think this comment and the one below should be reversed. We set output zero_point to 0 for unsigned types, right?

supriyar commented:
LGTM, do we have some perf numbers as well for this?
I wonder if it would be possible to get some benchmark numbers on mobile as well for the elementwise operators you've implemented so far. Vec256 wouldn't give any speedup on mobile and it would be nice to track what the perf is on mobile as well for future optimizations.

jamesr66a commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -306,6 +306,69 @@ void qsigmoid_kernel(const Tensor& qx, Tensor& qy) {
+        auto x = at::dequantize_val(scale, zero_point, qx);
+        x = x + 3.0f;
+        x = x > 0.0f ? x : 0.0f;
+        auto y = (x < 6.0f ? x : 6.0f) / 6.0f;
```
Same comment as the other PR, can do this branch free

vkuzo committed:
Update on "adds quantized implementation of hard sigmoid"


Summary:

Adds quantized version of hardsigmoid activation.

Note: not implementing the _ and .out versions is
currently intended, because the implementation changes the scale and
zp and it's nice to not allow the user to specify scale
and zp.  Lmk if we should handle this differently.

Test Plan:

tests
benchmarks

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
> LGTM, do we have some perf numbers as well for this?
> I wonder if it would be possible to get some benchmark numbers on mobile as well for the elementwise operators you've implemented so far. Vec256 wouldn't give any speedup on mobile and it would be nice to track what the perf is on mobile as well for future optimizations.

sure thing, here is the output of the current benchmark: https://our.internmc.facebook.com/intern/paste/P127415136/

I will follow up offline on the best way to benchmark all of these things on mobile!

supriyar reviewed (approved):
Thanks for the benchmarks! I usually also compare against float to see the speedup. 
But looks good overall

vkuzo committed:
Update on "adds quantized implementation of hard sigmoid"


Summary:

Adds quantized version of hardsigmoid activation.

Note: not implementing the _ and .out versions is
currently intended, because the implementation changes the scale and
zp and it's nice to not allow the user to specify scale
and zp.  Lmk if we should handle this differently.

Test Plan:

tests
benchmarks

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "adds quantized implementation of hard sigmoid"


Summary:

Adds quantized version of hardsigmoid activation.

Note: not implementing the _ and .out versions is
currently intended, because the implementation changes the scale and
zp and it's nice to not allow the user to specify scale
and zp.  Lmk if we should handle this differently.

Test Plan:

tests
benchmarks

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20480546](https://our.internmc.facebook.com/intern/diff/D20480546)

[ghstack-poisoned]

vkuzo committed:
Update on "adds quantized implementation of hard sigmoid"


Summary:

Adds quantized version of hardsigmoid activation.

Note: not implementing the _ and .out versions is
currently intended, because the implementation changes the scale and
zp and it's nice to not allow the user to specify scale
and zp.  Lmk if we should handle this differently.

Test Plan:

tests
benchmarks

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20480546](https://our.internmc.facebook.com/intern/diff/D20480546)

[ghstack-poisoned]

mrshenli commented:
Looks like this PR breaks master:

```
Mar 18 00:56:58 ======================================================================
Mar 18 00:56:58 FAIL [0.974s]: test_qhardsigmoid (__main__.TestQuantizedOps)
Mar 18 00:56:58 ----------------------------------------------------------------------
Mar 18 00:56:58 Traceback (most recent call last):
Mar 18 00:56:58   File "test_quantized.py", line 244, in test_qhardsigmoid
Mar 18 00:56:58     elements=hu.floats(-1e3, 1e3, allow_nan=False, allow_infinity=False),
Mar 18 00:56:58   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/hypothesis/core.py", line 1116, in wrapped_test
Mar 18 00:56:58     raise the_error_hypothesis_found
Mar 18 00:56:58   File "test_quantized.py", line 271, in test_qhardsigmoid
Mar 18 00:56:58     message="Hardsigmoid failed: {} vs. {}".format(qY, qY_hat))
Mar 18 00:56:58   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 880, in assertEqual
Mar 18 00:56:58     exact_dtype=exact_dtype)
Mar 18 00:56:58   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 882, in assertEqual
Mar 18 00:56:58     assertTensorsEqual(x, y)
Mar 18 00:56:58   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 848, in assertTensorsEqual
Mar 18 00:56:58     self.assertLessEqual(max_err, prec, message)
Mar 18 00:56:58 AssertionError: tensor(1, dtype=torch.int32) not less than or equal to 1e-05 : Hardsigmoid failed: tensor([0.5000], size=(1,), dtype=torch.qint32,
Mar 18 00:56:58        quantization_scheme=torch.per_tensor_affine, scale=2.3283064365386963e-10,
Mar 18 00:56:58        zero_point=0) vs. tensor([-0.5000], size=(1,), dtype=torch.qint32,
Mar 18 00:56:58        quantization_scheme=torch.per_tensor_affine, scale=2.3283064365386963e-10,
Mar 18 00:56:58        zero_point=0)
```

e.g.:

https://app.circleci.com/pipelines/github/pytorch/pytorch/141312/workflows/9036f70f-a41d-4968-aedc-4cab6b4307f0/jobs/4849783
https://app.circleci.com/pipelines/github/pytorch/pytorch/141312/workflows/9036f70f-a41d-4968-aedc-4cab6b4307f0/jobs/4849602

mrshenli commented:
@vkuzo I am reverting this PR to fix test signals on master. Please resubmit with a fix. Thanks!

vkuzo commented:
> @vkuzo I am reverting this PR to fix test signals on master. Please resubmit with a fix. Thanks!

sorry about that, will do!
